doc/generated/rpmem_persist.3
.\" Automatically generated by Pandoc 2.0.6
.\"
.TH "RPMEM_PERSIST" "3" "2020-01-16" "PMDK - rpmem API version 1.3" "PMDK Programmer's Manual"
.hy
.\" Copyright 2014-2020, Intel Corporation
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\"     * Redistributions of source code must retain the above copyright
.\"       notice, this list of conditions and the following disclaimer.
.\"
.\"     * Redistributions in binary form must reproduce the above copyright
.\"       notice, this list of conditions and the following disclaimer in
.\"       the documentation and/or other materials provided with the
.\"       distribution.
.\"
.\"     * Neither the name of the copyright holder nor the names of its
.\"       contributors may be used to endorse or promote products derived
.\"       from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.SH NAME
.PP
\f[B]rpmem_persist\f[](), \f[B]rpmem_deep_persist\f[](),
\f[B]rpmem_flush\f[](), \f[B]rpmem_drain\f[](), \f[B]rpmem_read\f[]() \-
functions to copy and read remote pools
.SH SYNOPSIS
.IP
.nf
\f[C]
#include\ <librpmem.h>

int\ rpmem_persist(RPMEMpool\ *rpp,\ size_t\ offset,
\ \ \ \ size_t\ length,\ unsigned\ lane,\ unsigned\ flags);
int\ rpmem_deep_persist(RPMEMpool\ *rpp,\ size_t\ offset,
\ \ \ \ size_t\ length,\ unsigned\ lane);

int\ rpmem_flush(RPMEMpool\ *rpp,\ size_t\ offset,
\ \ \ \ size_t\ length,\ unsigned\ lane,\ unsigned\ flags);
int\ rpmem_drain(RPMEMpool\ *rpp,\ unsigned\ lane,\ unsigned\ flags);

int\ rpmem_read(RPMEMpool\ *rpp,\ void\ *buff,\ size_t\ offset,
\ \ \ \ size_t\ length,\ unsigned\ lane);
\f[]
.fi
.SH DESCRIPTION
.PP
The \f[B]rpmem_persist\f[]() function copies data of given
\f[I]length\f[] at given \f[I]offset\f[] from the associated local
memory pool and makes sure the data is persistent on the remote node
before the function returns.
The remote node is identified by the \f[I]rpp\f[] handle which must be
returned from either \f[B]rpmem_open\f[](3) or \f[B]rpmem_create\f[](3).
The \f[I]offset\f[] is relative to the \f[I]pool_addr\f[] specified in
the \f[B]rpmem_open\f[](3) or \f[B]rpmem_create\f[](3) call.
If the remote pool was created using \f[B]rpmem_create\f[]() with
non\-NULL \f[I]create_attr\f[] argument, \f[I]offset\f[] has to be
greater or equal to 4096.
In that case the first 4096 bytes of the pool is used for storing the
pool metadata and cannot be overwritten.
If the pool was created with NULL \f[I]create_attr\f[] argument, the
pool metadata is not stored with the pool and \f[I]offset\f[] can be any
nonnegative number.
The \f[I]offset\f[] and \f[I]length\f[] combined must not exceed the
\f[I]pool_size\f[] passed to \f[B]rpmem_open\f[](3) or
\f[B]rpmem_create\f[](3).
The \f[B]rpmem_persist\f[]() operation is performed using the given
\f[I]lane\f[] number.
The lane must be less than the value returned by \f[B]rpmem_open\f[](3)
or \f[B]rpmem_create\f[](3) through the \f[I]nlanes\f[] argument (so it
can take a value from 0 to \f[I]nlanes\f[] \- 1).
The \f[I]flags\f[] argument can be 0 or RPMEM_PERSIST_RELAXED which
means the persist operation will be done without any guarantees
regarding atomicity of memory transfer.
.PP
The \f[B]rpmem_deep_persist\f[]() function works in the same way as
\f[B]rpmem_persist\f[](3) function, but additionally it flushes the data
to the lowest possible persistency domain available from software.
Please see \f[B]pmem_deep_persist\f[](3) for details.
.PP
The \f[B]rpmem_flush\f[]() and \f[B]rpmem_drain\f[]() functions are two
halves of the single \f[B]rpmem_persist\f[]().
The \f[B]rpmem_persist\f[]() copies data and makes it persistent in the
one shot, where \f[B]rpmem_flush\f[]() and \f[B]rpmem_drain\f[]() split
this operation into two stages.
The \f[B]rpmem_flush\f[]() copies data of given \f[I]length\f[] at a
given \f[I]offset\f[] from the associated local memory pool to the
remote node.
The \f[B]rpmem_drain\f[]() makes sure the data copied in all preceding
\f[B]rpmem_flush\f[]() calls is persistent on the remote node before the
function returns.
Data copied using \f[B]rpmem_flush\f[]() can not be considered
persistent on the remote node before return from following
\f[B]rpmem_drain\f[]().
Single \f[B]rpmem_drain\f[]() confirms persistence on the remote node of
data copied by all \f[B]rpmem_flush\f[]() functions called before it and
using the same \f[I]lane\f[].
The last \f[B]rpmem_flush\f[]() + \f[B]rpmem_drain\f[]() can be replaced
with \f[B]rpmem_persist\f[]() at no cost.
.PP
The \f[I]flags\f[] argument for \f[B]rpmem_flush\f[]() can be 0 or
RPMEM_FLUSH_RELAXED which means the flush operation will be done without
any guarantees regarding atomicity of memory transfer.
The \f[I]flags\f[] argument for \f[B]rpmem_drain\f[]() must be 0.
.PP
The \f[B]rpmem_flush\f[]() function performance is affected by
\f[B]RPMEM_WORK_QUEUE_SIZE\f[] environment variable (see
\f[B]librpmem\f[](7) for more details).
.PP
The \f[B]rpmem_read\f[]() function reads \f[I]length\f[] bytes of data
from a remote pool at \f[I]offset\f[] and copies it to the buffer
\f[I]buff\f[].
The operation is performed on the specified \f[I]lane\f[].
The lane must be less than the value returned by \f[B]rpmem_open\f[](3)
or \f[B]rpmem_create\f[](3) through the \f[I]nlanes\f[] argument (so it
can take a value from 0 to \f[I]nlanes\f[] \- 1).
The \f[I]rpp\f[] must point to a remote pool opened or created
previously by \f[B]rpmem_open\f[](3) or \f[B]rpmem_create\f[](3).
.SH RETURN VALUE
.PP
The \f[B]rpmem_persist\f[]() function returns 0 if the entire memory
area was made persistent on the remote node.
Otherwise it returns a non\-zero value and sets \f[I]errno\f[]
appropriately.
.PP
The \f[B]rpmem_flush\f[]() function returns 0 if duplication of the
memory area to the remote node was initialized successfully.
Otherwise, it returns a non\-zero value and sets \f[I]errno\f[]
appropriately.
.PP
The \f[B]rpmem_drain\f[]() function returns 0 if the memory areas
duplicated by all \f[B]rpmem_flush\f[]() calls preceding the
\f[B]rpmem_drain\f[]() are made persistent on the remote node.
Otherwise, it returns a non\-zero value and sets \f[I]errno\f[]
appropriately.
.PP
The \f[B]rpmem_read\f[]() function returns 0 if the data was read
entirely.
Otherwise it returns a non\-zero value and sets \f[I]errno\f[]
appropriately.
.SH CAVEATS
.PP
Ordering of \f[B]rpmem_flush\f[]() and \f[B]rpmem_persist\f[]()
operations which are using different \f[I]lane\f[] values is not
guaranteed.
.SH SEE ALSO
.PP
\f[B]rpmem_create\f[](3), \f[B]rpmem_open\f[](3),
\f[B]rpmem_persist\f[](3), \f[B]sysconf\f[](3), \f[B]limits.conf\f[](5),
\f[B]libpmemobj\f[](7) and \f[B]<http://pmem.io>\f[]


ChangeLog
Fri Jan 17 2020 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.8-rc1

	This release:
	- Introduces new API in libpmemobj for user-assisted defragmentation
	  (see pmemobj_defrag man page for details).
	- Introduces experimental support for PPC64LE architecture.
	- Introduces new flag in libpmemobj for opting-out of transaction aborts
	  on failure (POBJ_TX_NO_ABORT), along with new variants of existing APIs
	  that didn't accept flags (pmemobj_tx_xfree, pmemobj_tx_xpublish,
	  pmemobj_tx_xlock, pmemobj_tx_xlog_append_buffer, pmemobj_tx_xstrdup,
	  pmemobj_tx_xwcsdup).
	- Moves out libvmem and libvmmalloc to the new repository
	  (https://github.com/pmem/vmem).

	Other changes:
	- obj: introduce new statistics useful for defrag
	- obj: introduce transient/persistent statistics enabling
	- obj: introduce pmemobj_tx_(set/get)_user_data funcs
	- obj: introduce pmemobj_(set/get)_user_data funcs
	- obj: disable workaround for offsetof() since VS 15.5 in public header
	- common: drop support for libndctl < 63 on Linux
	- pool: rename -N --no-exec to -d --dry-run

	Notable bug fixes:
	- obj: fix potential NULL-dereference in ulog_store
	- obj: fix unintended recursive locking during palloc
	- obj: fix lock release order in palloc publish
	- obj: fix transient redo log of more than 64 entries
	- obj: fix capacity ulog calculation
	- obj: fix check of unaligned capacity size
	- rpmem: add a missing case for GPSPM + FLUSH_STRICT
	- pmem: fix pmemcheck support on platforms with eADR
	- pool: fix possible memory leak
	- rpmem: fix possible memory leak in rpmemd_config_read
	- rpmem: fix possible memory leak in rpmemd_log_init
	- rpmem: fix possible use-after-free

Mon Sep 30 2019 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.7

	This release:
	- Introduces new APIs in libpmemobj for managing space used by transactions.
	  (see pmemobj_tx_log_append_buffer man page for details)
	- Introduces new APIs in librpmem, splitting rpmem_persist into rpmem_flush
	  and rpmem_drain, allowing applications to use the flush + drain model
	  already known from libpmem. (libpmemobj does not use this feature yet)
	- Optimizes large libpmemobj transactions by significantly reducing
	  the amount of memory modified at the commit phase.
	- Optimizes tracking of libpmemobj reservations.
	- Adds new flags for libpmemobj's pmemobj_tx_xadd_range[_direct] API:
	  POBJ_XADD_NO_SNAPSHOT and POBJ_XADD_ASSUME_INITIALIZED, allowing
	  applications to optimize how memory is tracked by the library.

	To support some of the above changes the libpmemobj on-media layout had
	to be changed, which means that old pools have to be converted using
	pmdk-convert >= 1.7.

	Other changes:
	- obj: fix merging of ranges when NOFLUSH flag is used (#1100)
	- rpmem: fix closing of ssh connection (#995, #1060)
	- obj: abort transaction on pmemobj_tx_publish failure

	Internal changes:
	- test: fault injection tests for pmemblk, pmemlog, and pmemobj
	- test: improved Python testing framework
	- test: support real pmem in bad blocks tests
	- common: allow not building examples and benchmarks

Tue Aug 27 2019 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.6.1

	This release fixes possible pool corruptions on Windows
	(see https://github.com/pmem/pmdk/pull/3728 for details),
	improves compatibility with newer Linux kernels with
	respect to Device DAX detection, fixes pmemobj space management
	for large pools, improves compatibility with newer toolchains,
	incorporates build fixes for FreeBSD and fixes a number of
	smaller bugs.

	Detailed list of bug fixes:
	- common: (win) fix possible pool file coruption (#972, #715, #603)
	- common: implement correct / robust device_dax_alignment (#1071)
	- obj: fix recycler not locating unused chunks
	- doc: update pmemobj_tx_lock documentation wrt behavior on fail
	- common: fix persistent domain detection (#1093)
	- common: vecq: fix a pointer-to-struct aliasing violation
	  (crash on arm64)
	- common: fix minor issues related to ndctl linking
	- obj: drop recursion from pmemobj_next
	- common: fix bug in badblock file error handling
	- obj: fix handling of malloc failures
	- common: fix handling of malloc failures (ctl)
	- jemalloc: fix build with gcc 9
	- obj: don't overwrite errno when palloc_heap_check_remote fails
	- doc: fix pmreorder emit log macro
	- rpmem: change order of rpmem init (#1043)
	- common: Fix build failure due to unused macro PAGE_SIZE
	- common: support older versions of pkg-config
	- tools: link with release variant of pmemcommon
	- common: add PMDK prefix to local copy of queue.h (#990)
	- rpmem: switch to using an open coded basename (FreeBSD)
	- common: posix_fallocate: guard against integer underflow in
	  check (FreeBSD)
	- test: support Valgrind 3.15
	- test: skip if fi_info is missing
	- test: (win) fix sparsefile error handling
	- test: fix libpmempool_feature tests that match logs
	- test: remove vmem_delete test (#1074)
	- test: adjust matchfiles in vmem_valgrind_region test (#1087)
	- test: remove old log files for windows (#1013)
	- test: remove invalid expect_normal_exit (#1092)
	- test: suppress ld leak (#1098)
	- test: Expose necessary symbols in libvmmalloc_dummy_funcs
	  (FreeBSD)
	- test: fix tests failing because `tput` fails (FreeBSD)
	- test: avoid obj_critnib_mt taking very long on many-core machines
	- test: deal with libndctl's path without build system
	- test: overwrite old log in pmempool_create/TEST14.PS1
	- test: fix match files in tests which use dax devices
	- test: fix match file in rpmem_addr_ext test
	- test: fix pmempool_check test

Wed Aug 28 2019 Marcin Ślusarz <marcin.slusarz@intel.com>

	* Version 1.5.2

	This release fixes possible pool corruptions on Windows
	(see https://github.com/pmem/pmdk/pull/3728 for details),
	improves compatibility with newer Linux kernels with
	respect to Device DAX detection, fixes pmemobj space management
	for large pools, improves compatibility with newer toolchains

src/benchmarks/rpmem_persist.cpp
/*
 * Copyright 2016-2019, Intel Corporation
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *      * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *      * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in
 *       the documentation and/or other materials provided with the
 *       distribution.
 *
 *      * Neither the name of the copyright holder nor the names of its
 *        contributors may be used to endorse or promote products derived
 *        from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * rpmem_persist.cpp -- rpmem persist benchmarks definition
 */

#include <cassert>
#include <cerrno>
#include <cstddef>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <fcntl.h>
#include <sys/file.h>
#include <sys/mman.h>
#include <unistd.h>

#include "benchmark.hpp"
#include "libpmem.h"
#include "librpmem.h"
#include "microbench.h"
#include "os.h"
#include "set.h"
#include "util.h"

#define CL_ALIGNMENT 64
#define MAX_OFFSET (CL_ALIGNMENT - 1)

#define ALIGN_CL(x) (((x) + CL_ALIGNMENT - 1) & ~(CL_ALIGNMENT - 1))

/*
 * rpmem_args -- benchmark specific command line options
 */
struct rpmem_args {
	char *mode;	/* operation mode: stat, seq, rand */
	bool no_warmup;    /* do not do warmup */
	bool no_memset;    /* do not call memset before each persist */
	size_t chunk_size; /* elementary chunk size */
	size_t dest_off;   /* destination address offset */
	bool relaxed;      /* use RPMEM_PERSIST_RELAXED flag */
};

/*
 * rpmem_bench -- benchmark context
 */
struct rpmem_bench {
	struct rpmem_args *pargs; /* benchmark specific arguments */
	size_t *offsets;	  /* random/sequential address offsets */
	size_t n_offsets;	 /* number of random elements */
	int const_b;		  /* memset() value */
	size_t min_size;	  /* minimum file size */
	void *addrp;		  /* mapped file address */
	void *pool;		  /* memory pool address */
	size_t pool_size;	 /* size of memory pool */
	size_t mapped_len;	/* mapped length */
	RPMEMpool **rpp;	  /* rpmem pool pointers */
	unsigned *nlanes;	 /* number of lanes for each remote replica */
	unsigned nreplicas;       /* number of remote replicas */
	size_t csize_align;       /* aligned elementary chunk size */
	unsigned flags;		  /* flags for rpmem_persist */
};

/*
 * operation_mode -- mode of operation
 */
enum operation_mode {
	OP_MODE_UNKNOWN,
	OP_MODE_STAT,      /* always use the same chunk */
	OP_MODE_SEQ,       /* use consecutive chunks */
	OP_MODE_RAND,      /* use random chunks */
	OP_MODE_SEQ_WRAP,  /* use consequtive chunks, but use file size */
	OP_MODE_RAND_WRAP, /* use random chunks, but use file size */
};

/*
 * parse_op_mode -- parse operation mode from string
 */
static enum operation_mode
parse_op_mode(const char *arg)
{
	if (strcmp(arg, "stat") == 0)
		return OP_MODE_STAT;
	else if (strcmp(arg, "seq") == 0)
		return OP_MODE_SEQ;
	else if (strcmp(arg, "rand") == 0)
		return OP_MODE_RAND;
	else if (strcmp(arg, "seq-wrap") == 0)
		return OP_MODE_SEQ_WRAP;
	else if (strcmp(arg, "rand-wrap") == 0)
		return OP_MODE_RAND_WRAP;
	else
		return OP_MODE_UNKNOWN;
}

/*
 * init_offsets -- initialize offsets[] array depending on the selected mode
 */
static int
init_offsets(struct benchmark_args *args, struct rpmem_bench *mb,
	     enum operation_mode op_mode)
{
	size_t n_ops_by_size = (mb->pool_size - POOL_HDR_SIZE) /
		(args->n_threads * mb->csize_align);

	mb->n_offsets = args->n_ops_per_thread * args->n_threads;
	mb->offsets = (size_t *)malloc(mb->n_offsets * sizeof(*mb->offsets));
	if (!mb->offsets) {
		perror("malloc");
		return -1;
	}

	unsigned seed = args->seed;

	for (size_t i = 0; i < args->n_threads; i++) {
		for (size_t j = 0; j < args->n_ops_per_thread; j++) {
			size_t off_idx = i * args->n_ops_per_thread + j;
			size_t chunk_idx;
			switch (op_mode) {
				case OP_MODE_STAT:
					chunk_idx = i;
					break;
				case OP_MODE_SEQ:
					chunk_idx =
						i * args->n_ops_per_thread + j;
					break;
				case OP_MODE_RAND:
					chunk_idx = i * args->n_ops_per_thread +
						os_rand_r(&seed) %
							args->n_ops_per_thread;
					break;
				case OP_MODE_SEQ_WRAP:
					chunk_idx = i * n_ops_by_size +
						j % n_ops_by_size;
					break;
				case OP_MODE_RAND_WRAP:
					chunk_idx = i * n_ops_by_size +
						os_rand_r(&seed) %
							n_ops_by_size;
					break;
				default:
					assert(0);
					return -1;
			}

			mb->offsets[off_idx] = POOL_HDR_SIZE +
				chunk_idx * mb->csize_align +
				mb->pargs->dest_off;
		}
	}

	return 0;
}

/*
 * do_warmup -- does the warmup by writing the whole pool area
 */
static int
do_warmup(struct rpmem_bench *mb)
{
	/* clear the entire pool */
	memset((char *)mb->pool + POOL_HDR_SIZE, 0,
	       mb->pool_size - POOL_HDR_SIZE);

	for (unsigned r = 0; r < mb->nreplicas; ++r) {
		int ret = rpmem_persist(mb->rpp[r], POOL_HDR_SIZE,
					mb->pool_size - POOL_HDR_SIZE, 0,
					mb->flags);
		if (ret)
			return ret;
	}

	/* if no memset for each operation, do one big memset */
	if (mb->pargs->no_memset) {
		memset((char *)mb->pool + POOL_HDR_SIZE, 0xFF,
		       mb->pool_size - POOL_HDR_SIZE);
	}

	return 0;
}

/*
 * rpmem_op -- actual benchmark operation
 */
static int
rpmem_op(struct benchmark *bench, struct operation_info *info)
{
	auto *mb = (struct rpmem_bench *)pmembench_get_priv(bench);

	uint64_t idx = info->worker->index * info->args->n_ops_per_thread +
		info->index;

	assert(idx < mb->n_offsets);

	size_t offset = mb->offsets[idx];
	size_t len = mb->pargs->chunk_size;

	if (!mb->pargs->no_memset) {
		MBENCH_START("pmembench_rpmem_memset")
		{
			void *dest = (char *)mb->pool + offset;
			/*
			 * thread id on MS 4 bits and operation id on LS 4 bits
			 */
			int c = ((info->worker->index & 0xf) << 4) +
				((0xf & info->index));
			memset(dest, c, len);
#ifdef MBENCH_ENABLED
			MBENCH_IF
			{
				MBENCH_INC(idx);
				offset = mb->offsets[idx];
			}
#endif
		}
		MBENCH_STOP;
	}

	int ret = 0;
	MBENCH_START("pmembench_rpmem_persist")
	{
		for (unsigned r = 0; r < mb->nreplicas; ++r) {
			assert(info->worker->index < mb->nlanes[r]);

			ret = rpmem_persist(mb->rpp[r], offset, len,
					    info->worker->index, mb->flags);
			if (ret) {
				fprintf(stderr,
					"rpmem_persist replica #%u: %s\n", r,
					rpmem_errormsg());
				return ret;
			}
		}
#ifdef MBENCH_ENABLED
		MBENCH_IF
		{
			MBENCH_INC(idx);
			offset = mb->offsets[idx];
		}
#endif
	}
	MBENCH_STOP;

	return 0;
}

/*
 * rpmem_map_file -- map local file
 */
static int
rpmem_map_file(const char *path, struct rpmem_bench *mb, size_t size)
{
	int mode;
#ifndef _WIN32
	mode = S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH;
#else
	mode = S_IWRITE | S_IREAD;
#endif

	mb->addrp = pmem_map_file(path, size, PMEM_FILE_CREATE, mode,
				  &mb->mapped_len, nullptr);

	if (!mb->addrp)
		return -1;

	return 0;
}

/*
 * rpmem_unmap_file -- unmap local file
 */
static int
rpmem_unmap_file(struct rpmem_bench *mb)
{
	return pmem_unmap(mb->addrp, mb->mapped_len);
}

/*
 * rpmem_poolset_init -- read poolset file and initialize benchmark accordingly
 */
static int
rpmem_poolset_init(const char *path, struct rpmem_bench *mb,
		   struct benchmark_args *args)
{
	struct pool_set *set;
	struct pool_replica *rep;
	struct remote_replica *remote;
	struct pool_set_part *part;

	struct rpmem_pool_attr attr;
	memset(&attr, 0, sizeof(attr));
	memcpy(attr.signature, "PMEMBNCH", sizeof(attr.signature));

	/* read and validate poolset */
	if (util_poolset_read(&set, path)) {
		fprintf(stderr, "Invalid poolset file '%s'\n", path);
		return -1;
	}

	assert(set);
	if (set->nreplicas < 2) {
		fprintf(stderr, "No replicas defined\n");
		goto err_poolset_free;
	}

	if (set->remote == 0) {
		fprintf(stderr, "No remote replicas defined\n");
		goto err_poolset_free;
	}

	for (unsigned i = 1; i < set->nreplicas; ++i) {
		if (!set->replica[i]->remote) {
			fprintf(stderr, "Local replicas are not supported\n");
			goto err_poolset_free;
		}
	}

	/* read and validate master replica */
	rep = set->replica[0];

	assert(rep);
	assert(rep->remote == nullptr);
	if (rep->nparts != 1) {
		fprintf(stderr,
			"Multipart master replicas are not supported\n");
		goto err_poolset_free;
	}

	if (rep->repsize < mb->min_size) {
		fprintf(stderr, "A master replica is too small (%zu < %zu)\n",
			rep->repsize, mb->min_size);
		goto err_poolset_free;
	}

	part = (struct pool_set_part *)&rep->part[0];
	if (rpmem_map_file(part->path, mb, rep->repsize)) {
		perror(part->path);
		goto err_poolset_free;
	}

	mb->pool_size = mb->mapped_len;
	mb->pool = (void *)((uintptr_t)mb->addrp);

	/* prepare remote replicas */
	mb->nreplicas = set->nreplicas - 1;
	mb->nlanes = (unsigned *)malloc(mb->nreplicas * sizeof(unsigned));
	if (mb->nlanes == nullptr) {
		perror("malloc");
		goto err_unmap_file;
	}

	mb->rpp = (RPMEMpool **)malloc(mb->nreplicas * sizeof(RPMEMpool *));
	if (mb->rpp == nullptr) {
		perror("malloc");
		goto err_free_lanes;
	}

	unsigned r;
	for (r = 0; r < mb->nreplicas; ++r) {
		remote = set->replica[r + 1]->remote;

		assert(remote);

		mb->nlanes[r] = args->n_threads;
		/* Temporary WA for librpmem issue */
		++mb->nlanes[r];

		mb->rpp[r] = rpmem_create(remote->node_addr, remote->pool_desc,
					  mb->addrp, mb->pool_size,
					  &mb->nlanes[r], &attr);
		if (!mb->rpp[r]) {
			perror("rpmem_create");
			goto err_rpmem_close;
		}

		if (mb->nlanes[r] < args->n_threads) {
			fprintf(stderr,
				"Number of threads too large for replica #%u (max: %u)\n",
				r, mb->nlanes[r]);
			r++; /* close current replica */
			goto err_rpmem_close;
		}
	}

	util_poolset_free(set);
	return 0;

err_rpmem_close:
	for (unsigned i = 0; i < r; i++)
		rpmem_close(mb->rpp[i]);
	free(mb->rpp);

err_free_lanes:
	free(mb->nlanes);

err_unmap_file:
	rpmem_unmap_file(mb);

err_poolset_free:
	util_poolset_free(set);
	return -1;
}

/*
 * rpmem_poolset_fini -- close opened local and remote replicas
 */
static void
rpmem_poolset_fini(struct rpmem_bench *mb)
{
	for (unsigned r = 0; r < mb->nreplicas; ++r) {
		rpmem_close(mb->rpp[r]);
	}

	rpmem_unmap_file(mb);
}

/*
 * rpmem_set_min_size -- compute minimal file size based on benchmark arguments
 */
static void
rpmem_set_min_size(struct rpmem_bench *mb, enum operation_mode op_mode,
		   struct benchmark_args *args)
{
	mb->csize_align = ALIGN_CL(mb->pargs->chunk_size);

	if (MBENCH_PROBE_STRSTR("pmembench_rpmem_")) {
		/*
		 * if micro-benchmark is enabled alter parameters according to
		 * micro-benchmark requirements
		 */
		args->n_ops_per_thread = MBENCH_GET_REPEAT;
	}

	switch (op_mode) {
		case OP_MODE_STAT:
			mb->min_size = mb->csize_align * args->n_threads;
			break;
		case OP_MODE_SEQ:
		case OP_MODE_RAND:
			mb->min_size = mb->csize_align *
				args->n_ops_per_thread * args->n_threads;
			break;
		case OP_MODE_SEQ_WRAP:
		case OP_MODE_RAND_WRAP:
			/*
			 * at least one chunk per thread to avoid false sharing
			 */
			mb->min_size = mb->csize_align * args->n_threads;
			break;
		default:
			assert(0);
	}

	mb->min_size += POOL_HDR_SIZE;
}

/*
 * rpmem_init -- initialization function
 */
static int
rpmem_init(struct benchmark *bench, struct benchmark_args *args)
{
	assert(bench != nullptr);
	assert(args != nullptr);
	assert(args->opts != nullptr);

	MBENCH_INIT();

	auto *mb = (struct rpmem_bench *)malloc(sizeof(struct rpmem_bench));
	if (!mb) {
		perror("malloc");
		return -1;
	}

	mb->flags = 0;
	mb->pargs = (struct rpmem_args *)args->opts;
	mb->pargs->chunk_size = args->dsize;

	if (mb->pargs->relaxed)
		mb->flags |= RPMEM_PERSIST_RELAXED;

	enum operation_mode op_mode = parse_op_mode(mb->pargs->mode);
	if (op_mode == OP_MODE_UNKNOWN) {
		fprintf(stderr, "Invalid operation mode argument '%s'\n",
			mb->pargs->mode);
		goto err_parse_mode;
	}

	rpmem_set_min_size(mb, op_mode, args);

	if (rpmem_poolset_init(args->fname, mb, args)) {
		goto err_poolset_init;
	}

	/* initialize offsets[] array depending on benchmark args */
	if (init_offsets(args, mb, op_mode) < 0) {
		goto err_init_offsets;
	}

	if (!mb->pargs->no_warmup) {
		if (do_warmup(mb) != 0) {
			fprintf(stderr, "do_warmup() function failed.\n");
			goto err_warmup;
		}
	}

	pmembench_set_priv(bench, mb);

	return 0;
err_warmup:
	free(mb->offsets);
err_init_offsets:
	rpmem_poolset_fini(mb);
err_poolset_init:
err_parse_mode:
	free(mb);
	return -1;
}

/*
 * rpmem_exit -- benchmark cleanup function
 */
static int
rpmem_exit(struct benchmark *bench, struct benchmark_args *args)
{
	auto *mb = (struct rpmem_bench *)pmembench_get_priv(bench);
	rpmem_poolset_fini(mb);
	free(mb->offsets);
	free(mb);

	MBENCH_FINI();

	return 0;
}

static struct benchmark_clo rpmem_clo[5];
/* Stores information about benchmark. */
static struct benchmark_info rpmem_info;
CONSTRUCTOR(rpmem_persist_constructor)
void
pmem_rpmem_persist(void)
{

	rpmem_clo[0].opt_short = 'M';
	rpmem_clo[0].opt_long = "mem-mode";
	rpmem_clo[0].descr = "Memory writing mode :"
			     " stat, seq[-wrap], rand[-wrap]";
	rpmem_clo[0].def = "seq";
	rpmem_clo[0].off = clo_field_offset(struct rpmem_args, mode);
	rpmem_clo[0].type = CLO_TYPE_STR;

	rpmem_clo[1].opt_short = 'D';
	rpmem_clo[1].opt_long = "dest-offset";
	rpmem_clo[1].descr = "Destination cache line "
			     "alignment offset";
	rpmem_clo[1].def = "0";
	rpmem_clo[1].off = clo_field_offset(struct rpmem_args, dest_off);
	rpmem_clo[1].type = CLO_TYPE_UINT;
	rpmem_clo[1].type_uint.size =
		clo_field_size(struct rpmem_args, dest_off);
	rpmem_clo[1].type_uint.base = CLO_INT_BASE_DEC;
	rpmem_clo[1].type_uint.min = 0;
	rpmem_clo[1].type_uint.max = MAX_OFFSET;

	rpmem_clo[2].opt_short = 'w';
	rpmem_clo[2].opt_long = "no-warmup";
	rpmem_clo[2].descr = "Don't do warmup";
	rpmem_clo[2].def = "false";
	rpmem_clo[2].type = CLO_TYPE_FLAG;
	rpmem_clo[2].off = clo_field_offset(struct rpmem_args, no_warmup);

	rpmem_clo[3].opt_short = 'T';
	rpmem_clo[3].opt_long = "no-memset";
	rpmem_clo[3].descr = "Don't call memset for all rpmem_persist";
	rpmem_clo[3].def = "false";
	rpmem_clo[3].off = clo_field_offset(struct rpmem_args, no_memset);
	rpmem_clo[3].type = CLO_TYPE_FLAG;

	rpmem_clo[4].opt_short = 0;
	rpmem_clo[4].opt_long = "persist-relaxed";
	rpmem_clo[4].descr = "Use RPMEM_PERSIST_RELAXED flag";
	rpmem_clo[4].def = "false";
	rpmem_clo[4].off = clo_field_offset(struct rpmem_args, relaxed);
	rpmem_clo[4].type = CLO_TYPE_FLAG;

	rpmem_info.name = "rpmem_persist";
	rpmem_info.brief = "Benchmark for rpmem_persist() "
			   "operation";
	rpmem_info.init = rpmem_init;
	rpmem_info.exit = rpmem_exit;
	rpmem_info.multithread = true;
	rpmem_info.multiops = true;
	rpmem_info.operation = rpmem_op;
	rpmem_info.measure_time = true;
	rpmem_info.clos = rpmem_clo;
	rpmem_info.nclos = ARRAY_SIZE(rpmem_clo);
	rpmem_info.opts_size = sizeof(struct rpmem_args);
	rpmem_info.rm_file = true;
	rpmem_info.allow_poolset = true;
	rpmem_info.print_bandwidth = true;
	REGISTER_BENCHMARK(rpmem_info);
};
